lib/pacemaker/pcmk_sched_primitive.c

```diff
@@ -386,15 +389,48 @@ pcmk__primitive_assign(pe_resource_t *rsc, const pe_node_t *prefer)
     pe__show_node_weights(true, rsc, "Pre-assignment", rsc->allowed_nodes,
                           rsc->cluster);
 
-    colocations = pcmk__this_with_colocations(rsc);
-    g_list_foreach(colocations, apply_this_with, rsc);
-    g_list_free(colocations);
-    pe__show_node_weights(true, rsc, "Post-this-with", rsc->allowed_nodes,
-                          rsc->cluster);
+    this_with_colocations = pcmk__this_with_colocations(rsc);
+    with_this_colocations = pcmk__with_this_colocations(rsc);
+
+    // Apply mandatory colocations first, to satisfy as many as possible
+    for (iter = this_with_colocations; iter != NULL; iter = iter->next) {
+        colocation = iter->data;
+        if ((colocation->score <= -CRM_SCORE_INFINITY)
+            || (colocation->score >= CRM_SCORE_INFINITY)) {
+            apply_this_with(iter->data, rsc);
+        }
+    }
+    for (iter = with_this_colocations; iter != NULL; iter = iter->next) {
```

Comment:
This looks right, but of course previously we weren't considering "with this" colocations at all, regardless of mandatory vs. optional. Hopefully considering all mandatory constraints is correct. Considering "with this" colocations may even be necessary for the fix, although it's not obvious from the commit message.

No tests change materially.

---

On a related note: The regression test and T217 both deal only with optional anti-colocations. This block (and this PR) don't seem to change the behavior for mandatory anti-colocations. Perhaps we should add a test -- even if we don't want to change it, to codify how we think it should work.

Specifically, I'm thinking of the following situation, based on the `group-anticolocation` test:
* `group2` with `group1`: `-INFINITY`
* `group1` running on `node2`, `group2` running on `node1`
* `member2b` is failed and has reached migration threshold

We have two options:
1. Move `group1` to `node1` so that `group2` can then run all its resources on `node2`.
2. Leave `member2b` stopped due to the mandatory anti-colocation and to `group1`'s earlier placement.

Currently, we have behavior (2). Behavior (1) would allow more resources to run in a way that still adheres to the constraints, while behavior (2) avoids disrupting the primary resource.

Reply:
> This looks right, but of course previously we weren't considering "with this" colocations at all, regardless of mandatory vs. optional. Hopefully considering all mandatory constraints is correct. Considering "with this" colocations may even be necessary for the fix, although it's not obvious from the commit message.

Before, we considered all "this with" colocations, followed by all "with this" colocations (via apply_with_this(), now named pcmk__add_dependent_scores()). Now, we consider mandatory "this with" colocations, followed by mandatory "with this" colocations, followed by optional "this with" colocations, followed by optional "with this" colocations.

> No tests change materially.
> 
> On a related note: The regression test and T217 both deal only with optional anti-colocations. This block (and this PR) don't seem to change the behavior for mandatory anti-colocations. Perhaps we should add a test -- even if we don't want to change it, to codify how we think it should work.
> 
> Specifically, I'm thinking of the following situation, based on the `group-anticolocation` test:
> 
>     * `group2` with `group1`: `-INFINITY`
> 
>     * `group1` running on `node2`, `group2` running on `node1`
> 
>     * `member2b` is failed and has reached migration threshold
> 
> 
> We have two options:
> 
>     1. Move `group1` to `node1` so that `group2` can then run all its resources on `node2`.
> 
>     2. Leave `member2b` stopped due to the mandatory anti-colocation and to `group1`'s earlier placement.

Definitely 2. A mandatory colocation must always be respected.

> 
> Currently, we have behavior (2). Behavior (1) would allow more resources to run in a way that still adheres to the constraints, while behavior (2) avoids disrupting the primary resource.

Reply:
> Before, we considered all "this with" colocations, followed by all "with this" colocations (via apply_with_this(), now named pcmk__add_dependent_scores()). Now, we consider mandatory "this with" colocations, followed by mandatory "with this" colocations, followed by optional "this with" colocations, followed by optional "with this" colocations.

I'm sorry, this was a simple oversight on my part. The comparison view shoves the old "with this" consideration way down at the bottom, where the frees are now. I overlooked it.

This looks good to me.

> Definitely 2. A mandatory colocation must always be respected.

Well, it would be. The resources would never run on the same node. It's just a question of how much we value each of the following when assigning the primary resource in a mandatory anti-colocation:
* the primary's current location
* the (chained) preferences of its anti-colocated dependent resources

(1) would mean the end result changing from
```
    * Resource Group: group1:
      * member1a    (ocf:pacemaker:Dummy):   Started node2
      * member1b    (ocf:pacemaker:Dummy):   Started node2
    * Resource Group: group2:
      * member2a    (ocf:pacemaker:Dummy):   Started node1
      * member2b    (ocf:pacemaker:Dummy):   Stopped
```

to
```
    * Resource Group: group1:
      * member1a    (ocf:pacemaker:Dummy):   Started node1
      * member1b    (ocf:pacemaker:Dummy):   Started node1
    * Resource Group: group2:
      * member2a    (ocf:pacemaker:Dummy):   Started node2
      * member2b    (ocf:pacemaker:Dummy):   Started node2
```

The fact that `member2a` is Started might make things a little more complicated as well: `group2` is partially active.

It's fine with me if we stick with the current behavior -- it might also be less surprising to users. Just a scenario I thought about while looking at the updated test results.

Reply:
Hmm, that's a good question. Pacemaker does place a high priority on running as many resources as possible. Of course an infinite stickiness should keep the current behavior, but otherwise moving the resources around makes sense.

I'll look into that separately from this PR.

Reply:
Yeah it's tricky even aside from the implementation. Behavior (1) probably might've surprised me when I was on support, and infinite stickiness has broad reach -- a user might want the primary resource to be infinitely sticky only with regard to the dependent's preference or migration threshold. Things might also get weird for infinite stickiness in multiple/chained colocations.

I've given very little thought to the implications. They would definitely need heavy consideration.

Reply:
I suppose the problem is that R1 might move unnecessarily. If R2 *can't* run without moving R1, then moving R1 makes sense. But if R2 *can* still run (for example, to a third node in this scenario), we wouldn't want to move R1 just because R2 suddenly doesn't like some node.

Reply:
FYI, using a mandatory anti-colocation, and dropping the "double negative" check, the groups swap nodes if stickiness is finite and stay where they are if stickiness is infinite.

I'm thinking a good middle ground might be to consider double-negative preferences only if stickiness is zero. However at that point, we only have the dependent resource, and we'd want the primary's stickiness. :-/ It's probably doable, just painful.